compute_endpoint/globus_compute_endpoint/engines/high_throughput/engine.py

```diff
@@ -0,0 +1,924 @@
+        return msg["data"]
+
+    # When the executor gets lost, the weakref callback will wake up
+    # the queue management thread.
+    def weakref_cb(self, q=None):
+        """We do not use this yet."""
+        q.put(None)
+
+    def _start_queue_management_thread(self):
+        """Method to start the management thread as a daemon.
+
+        Checks if a thread already exists, then starts it.
+        Could be used later as a restart if the management thread dies.
+        """
+        if self._queue_management_thread is None:
+            log.debug("Starting queue management thread")
+            self._queue_management_thread = threading.Thread(
+                target=self._queue_management_worker, name="Queue-Management"
+            )
+            self._queue_management_thread.daemon = True
+            self._queue_management_thread.start()
+            log.debug("Started queue management thread")
+
+        else:
+            log.debug("Management thread already exists, returning")
+
+    def hold_worker(self, worker_id):
+        """Puts a worker on hold, preventing scheduling of additional tasks to it.
+
+        This is called "hold" mostly because this only stops scheduling of tasks,
+        and does not actually kill the worker.
+
+        Parameters
+        ----------
+
+        worker_id : str
+            Worker id to be put on hold
+        """
+        c = self.command_client.run(f"HOLD_WORKER;{worker_id}")
+        log.debug(f"Sent hold request to worker: {worker_id}")
+        return c
+
+    def send_heartbeat(self):
+        log.warning("Sending heartbeat to interchange")
+        msg = Heartbeat(endpoint_id="")
+        self.outgoing_q.put(msg.pack())
+
+    def wait_for_endpoint(self):
+        heartbeat = self.command_client.run(HeartbeatReq())
+        log.debug("Attempting heartbeat to interchange")
+        return heartbeat
+
+    @property
+    def outstanding(self):
+        outstanding_c = self.command_client.run("OUTSTANDING_C")
+        log.debug(f"Got outstanding count: {outstanding_c}")
+        return outstanding_c
+
+    @property
+    def connected_workers(self):
+        workers = self.command_client.run("MANAGERS")
+        log.debug(f"Got managers: {workers}")
+        return workers
+
+    def _submit(
+        self,
+        func: t.Callable,
+        *args: t.Any,
+        task_id: uuid.UUID = None,
+        **kwargs: t.Any,
+    ) -> Future:
+        self._task_counter += 1
+        if task_id is None:
+            s_task_id = str(self._task_counter)
+        else:
+            s_task_id = str(task_id)
+
+        ser_payload = fx_serializer.serialize((func, args, kwargs))
+        payload = Task(s_task_id, "RAW", ser_payload)
+        self.tasks[s_task_id] = HTEXFuture(self, task_id=s_task_id)
+        self.tasks[s_task_id].task_id = s_task_id
+        assert self.outgoing_q
+        self.outgoing_q.put(payload.pack())
+
+        return self.tasks[s_task_id]
```

Comment:
1. Is the second assignment of `.task_id` redundant?  Looks like the constructor has already been given the `task_id=s_task_id`.
2. Mild performance suggestion: can pay the dictionary lookup once if you keep a local reference to the HTEXFuture.  Consider:
```python
fut = HTEXFuture(self, task_id=s_task_id)
self.tasks[s_task_id] = fut  # "paid"; now just use fut
...
return fut
```